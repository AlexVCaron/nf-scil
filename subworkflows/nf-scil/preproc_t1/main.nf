// ** Importing modules from nf-scil ** //
include { DENOISING_NLMEANS } from '../../../modules/nf-scil/denoising/nlmeans/main'
include { PREPROC_N4 } from '../../../modules/nf-scil/preproc/n4/main'
include { IMAGE_RESAMPLE } from '../../../modules/nf-scil/image/resample/main'
include { BETCROP_ANTSBET } from '../../../modules/nf-scil/betcrop/antsbet/main'
include { BETCROP_CROPVOLUME } from '../../../modules/nf-scil/betcrop/cropvolume/main'

workflow PREPROC_T1 {

    take:
        ch_image           // channel: [ val(meta), [ image ] ]
        ch_template        // channel: [ val(meta), [ template ] ]
        ch_probability_map // channel: [ val(meta), [ probability_map ] ]

    main:

        ch_versions = Channel.empty()

        // ** Denoising ** //
        DENOISING_NLMEANS ( ch_image )
        ch_versions = ch_versions.mix(DENOISING_NLMEANS.out.versions.first())

        // ** N4 correction ** //
        ch_N4 = DENOISING_NLMEANS.out.image.map{[it[0], it[1],[],[]]}
        PREPROC_N4 ( ch_N4 )
        ch_versions = ch_versions.mix(PREPROC_N4.out.versions.first())

        // ** Resampling ** //
        ch_resampling = PREPROC_N4.out.image.map{[it[0], it[1],[]]}
        IMAGE_RESAMPLE ( ch_resampling )
        ch_versions = ch_versions.mix(IMAGE_RESAMPLE.out.versions.first())

        // ** Brain extraction ** //
        ch_betcop = IMAGE_RESAMPLE.out.image.join(ch_template).join(ch_probability_map)
        BETCROP_ANTSBET ( ch_betcop )
        ch_versions = ch_versions.mix(BETCROP_ANTSBET.out.versions.first())

        // ** crop ** //
        ch_crop = BETCROP_ANTSBET.out.t1.map{[it[0], it[1],[]]}
        BETCROP_CROPVOLUME ( ch_crop )
        ch_versions = ch_versions.mix(BETCROP_CROPVOLUME.out.versions.first())

    emit:
        image    = BETCROP_CROPVOLUME.out.image // channel: [ val(meta), [ image ] ]
        versions = ch_versions                  // channel: [ versions.yml ]
}

