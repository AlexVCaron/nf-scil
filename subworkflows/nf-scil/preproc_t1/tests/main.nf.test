nextflow_workflow {

    name "Test Subworkflow PREPROC_T1"
    script "../main.nf"
    workflow "PREPROC_T1"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/preproc_t1"

    tag "denoising/nlmeans"
    tag "preproc/n4"
    tag "image/resample"
    tag "betcrop/antsbet"
    tag "betcrop/cropvolume"

    tag "load_test_data"

        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "antsbet.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("preproc_t1") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_unaligned.nii.gz")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[2] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
