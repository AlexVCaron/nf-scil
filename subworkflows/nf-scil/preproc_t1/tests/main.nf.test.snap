{
    "preproc_t1": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b6d25b37b1101fea5c77bbbc711757c3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,ef0d9af46aa1e330c5d7679ead500596"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,88bae6bfeea1734111621b437e09548d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet.nii.gz:md5,016f12c75e3543c27987a91eb2160a19"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet_mask.nii.gz:md5,cf0cc416ca8b352d3cba58d9132ab0da"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,f4bea1286a90760eecf348dd9bffd2e6"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_bbox.pkl:md5,64485c09488a1dca9ae86826eaa4afb9"
                    ]
                ],
                "7": [
                    "versions.yml:md5,15044237a18a642d172b2bdb3abcad78",
                    "versions.yml:md5,34e64786c37df4e14d2ee9f57c93b0e3",
                    "versions.yml:md5,37327d900785469fe8aba4e90c045be3",
                    "versions.yml:md5,792251134d89d9f82a5c75d6d323ae0f",
                    "versions.yml:md5,b2baa0cca8ea5661b95a16dd3e8b45ca"
                ],
                "crop_box": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_bbox.pkl:md5,64485c09488a1dca9ae86826eaa4afb9"
                    ]
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_cropped.nii.gz:md5,f4bea1286a90760eecf348dd9bffd2e6"
                    ]
                ],
                "image_N4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__image_n4.nii.gz:md5,ef0d9af46aa1e330c5d7679ead500596"
                    ]
                ],
                "image_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet.nii.gz:md5,016f12c75e3543c27987a91eb2160a19"
                    ]
                ],
                "image_nlmeans": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b6d25b37b1101fea5c77bbbc711757c3"
                    ]
                ],
                "image_resample": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__resampled.nii.gz:md5,88bae6bfeea1734111621b437e09548d"
                    ]
                ],
                "mask_bet": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__t1_bet_mask.nii.gz:md5,cf0cc416ca8b352d3cba58d9132ab0da"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,15044237a18a642d172b2bdb3abcad78",
                    "versions.yml:md5,34e64786c37df4e14d2ee9f57c93b0e3",
                    "versions.yml:md5,37327d900785469fe8aba4e90c045be3",
                    "versions.yml:md5,792251134d89d9f82a5c75d6d323ae0f",
                    "versions.yml:md5,b2baa0cca8ea5661b95a16dd3e8b45ca"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-04-18T18:22:38.501159"
    }
}